docs/reference/fields.txt

```diff
@@ -576,93 +578,341 @@ you have three options:
+  Values <uncastable-values>` for more details.
 
-  .. code-block:: ruby
+Lastly, the class method ``evolve`` is similar to ``mongoize``, however it is used
+when transforming objects for use in Mongoid query criteria.
 
-    field :submitted_at, type: Time, default: Time.now
-    field :submitted_at, type: Time, default: ->{ Time.now }
+.. code-block:: ruby
 
-  The second definition is most likely the desired one, which causes the
-  time of submission to be set to the current time at the moment of
-  document instantiation.
+   point = Point.new(12, 24)
+   Venue.where(location: point) # This uses Point.evolve
 
-To set a default which depends on the document's state, use ``self``
-inside the ``Proc`` instance which would evaluate to the document instance
-being operated on:
+
+.. _phantom-custom-field-types:
+
+Phantom Custom Field Types
+--------------------------
+
+The custom field type may perform conversions from user-visible attribute
+values to the values stored in the database when the user-visible attribute
+value type is different from the declared field type. For example, this
+can be used to implement a mapping from one enumeration to another, to
+have more descriptive values in the application and more compact values stored
+in the database:
 
 .. code-block:: ruby
 
-  field :fulfill_by, type: Time, default: ->{
+  class ColorMapping
+
+    MAPPING = {
+      'black' => 0,
+      'white' => 1,
+    }.freeze
+
+    INVERSE_MAPPING = MAPPING.invert.freeze
+
+    class << self
+
+      # Takes application-scope value and converts it to how it would be
+      # stored in the database. Converts invalid values to nil.
+      def mongoize(object)
+        MAPPING[object]
+      end
+
+      # Get the value as it was stored in the database, and convert to
+      # application-scope value. Converts invalid values to nil.
+      def demongoize(object)
+        INVERSE_MAPPING[object]
+      end
+
+      # Converts the object that was supplied to a criteria and converts it
+      # into a query-friendly form. Returns invalid values as is.
+      def evolve(object)
+        MAPPING.fetch(object, object)
+      end
+    end
+  end
+
+  class Profile
+    include Mongoid::Document
+    field :color, type: ColorMapping
+  end
+
+  profile = Profile.new(color: 'white')
+  profile.color
+  # => "white"
+
+  # Writes 0 to color field
+  profile.save!
+
+
+.. _custom-field-options:
+
+Custom Field Options
+====================
+
+You may define custom options for the ``field`` macro function
+which extend its behavior at the your time model classes are loaded.
+
+As an example, we will define a ``:required`` option which will add a presence
+validator for the field. First, declare the new field option in an initializer,
+specifiying its handler function as a block:
+
+.. code-block:: ruby
+
+  # in /config/initializers/mongoid_custom_fields.rb
+
+  Mongoid::Fields.option :required do |model, field, value|
```

Comment:
`required` is already a defined option on fields.

Reply:
It's used here: https://github.com/mongodb/mongoid/blob/fe9d44baae7f9fb8c9ba561e3cb33561bdfe02ce/lib/mongoid/association/referenced/belongs_to.rb#L208

Reply:
Filed https://jira.mongodb.org/browse/MONGOID-5461 for this.